### PR TITLE
Implement union-based heaps: skew and pairing

### PR DESCRIPTION
**Pull request by kachayev**
Implementation for:

* skew heap

* pairing heap

Documentation is pure writing and has to be adjusted (TBD). Share it to hear feedback.

### CODE REVIEW

**Comment by vkostyukov** [comment]
Can we add some notes about asymptotic "running time" and "memory" in `PairingHeap`? Otherwise, looks pretty solid. Thanks @kachayev!

File: src/heap/PairingHeap.scala [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  /**
+   * Min value of this heap.
+   */
+  def min: A
+
+  /**
+   * Subtrees (child of this heap).
+   */
+  def childs: List[Heap[A]]
```
Comment (vkostyukov) [inline src/heap/PairingHeap.scala]: "children"?

**Comment by kachayev** [comment]
1. Added time/space complexity for Pairing Heap

2. Added basic explanation for most operations with Pairing Heaps

3. Fixed typo

**Comment by vkostyukov** [comment]
Nice! Thanks @kachayev!

**Pull request merged**